rollup/rollupsyncservice/rollupsyncservice.go

```diff
@@ -0,0 +1,419 @@
+	}
+
+	chunks, ok := decoded[2].([][]byte)
+	if !ok {
+		return nil, fmt.Errorf("failed to cast chunks to slice of byte slices")
+	}
+
+	return DecodeChunkBlockRanges(chunks)
+}
+
+// getBlocksInRange retrieves blocks from the blockchain within specified chunk ranges.
+func (s *RollupSyncService) getBlocksInRange(chunkRanges []*rawdb.ChunkBlockRange) ([]*types.Block, error) {
+	var blocks []*types.Block
+
+	latestBlockNumber := s.bc.CurrentBlock().Number().Uint64()
+
+	var maxRequestedBlockNumber uint64
+	for _, chunkRange := range chunkRanges {
+		if chunkRange.EndBlockNumber > maxRequestedBlockNumber {
+			maxRequestedBlockNumber = chunkRange.EndBlockNumber
+		}
+	}
+
+	if latestBlockNumber < maxRequestedBlockNumber {
+		time.Sleep(180 * time.Second)
+		return nil, fmt.Errorf("local node is not synced up to the required block height: %v", maxRequestedBlockNumber)
+	}
+
+	for _, chunkRange := range chunkRanges {
+		for i := chunkRange.StartBlockNumber; i <= chunkRange.EndBlockNumber; i++ {
+			block := s.bc.GetBlockByNumber(i)
+			if block == nil {
+				return nil, fmt.Errorf("failed to get block by number: %v", i)
+			}
+			blocks = append(blocks, block)
+		}
+	}
+
+	return blocks, nil
+}
+
+// convertBlocksToChunks processes and groups blocks into chunks based on the provided chunk ranges.
+func (s *RollupSyncService) convertBlocksToChunks(blocks []*types.Block, chunkRanges []*rawdb.ChunkBlockRange) ([]*Chunk, error) {
+	if len(blocks) == 0 {
+		return nil, fmt.Errorf("invalid arg: empty blocks")
+	}
+
+	wrappedBlocks := make([]*WrappedBlock, len(blocks))
+	for i, block := range blocks {
+		txData := txsToTxsData(block.Transactions())
+		state, err := s.bc.StateAt(block.Hash())
+		if err != nil {
+			return nil, fmt.Errorf("failed to get block state, block: %v, err: %w", block.Hash().Hex(), err)
+		}
+		withdrawRoot := withdrawtrie.ReadWTRSlot(rcfg.L2MessageQueueAddress, state)
+		wrappedBlocks[i] = &WrappedBlock{
+			Header:       block.Header(),
+			Transactions: txData,
+			WithdrawRoot: withdrawRoot,
+		}
+	}
+
+	minBlockNumber := blocks[0].Header().Number.Uint64()
+	var chunks []*Chunk
+	for _, cr := range chunkRanges {
+		start, end := cr.StartBlockNumber-minBlockNumber, cr.EndBlockNumber-minBlockNumber
+		// ensure start and end are within valid range.
+		if start < 0 || end >= uint64(len(wrappedBlocks)) || start > end {
+			return nil, fmt.Errorf("invalid chunk range, start: %v, end: %v, block len: %v", start, end, len(wrappedBlocks))
+		}
+		chunk := &Chunk{
+			Blocks: wrappedBlocks[start : end+1],
+		}
+		chunks = append(chunks, chunk)
+	}
+
+	return chunks, nil
+}
+
+// validateBatch verifies the consistency between l1 contract and l2 node data.
+// close the node and exit once any consistency check fails.
+func (s *RollupSyncService) validateBatch(batchIndex uint64, batchHash common.Hash, stateRoot common.Hash, withdrawRoot common.Hash, parentBatchMeta *rawdb.FinalizedBatchMeta, chunks []*Chunk) error {
```

Comment:
I think if we pass `L1FinalizeBatchEvent` into this function,
and then compare
+ localWithdrawRoot vs event.WithdrawRoot 
+ localStateRoot vs event.StateRoot 
+ localBatchHash vs event.BatchHash

the codes will be more easy to understand

Reply:
Yes. changed in [c08faf3](https://github.com/scroll-tech/go-ethereum/pull/515/commits/c08faf3d3a0b0d0c7704482d9ba5dce6c6e9fa30)